pkg/kubectl/cmd/apply.go

```diff
@@ -245,7 +245,13 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 	// Create the versioned struct from the type defined in the restmapping
 	// (which is the API version we'll be submitting the patch to)
 	versionedObject, err := api.Scheme.New(p.mapping.GroupVersionKind)
-	if err != nil {
+	switch {
+	case runtime.IsNotRegisteredError(err):
+		versionedObject, _, err = p.decoder.Decode(current, nil, nil)
```

Comment:
That sounds like a good starting point to fix this issue, though I'd like to see that expanded to take advantage of openapi schema info that can be retrieved from the server in the future